pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,328 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InfraProvider implements the infrastructure/provider interface
+// which is the interface between the rest of the installer and
+// the CAPI infrastructure provider.
+type InfraProvider struct {
+	infrastructure.Provider
+
+	capiProvider Provider
+}
+
+// InitializeProvider returns a CAPI provider implementation for a specific
+// cloud platform.
+func InitializeProvider(platform Provider) infrastructure.Provider {
+	return InfraProvider{capiProvider: platform}
+}
+
+// Provider provides an interface, intended for platform-specific implementations,
+// for calling functions at different points of the CAPI infrastructure provisioning
+// lifecycle.
+type Provider interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// Ignition generates the ignition secrets for bootstrap and control-plane machines
+	// and handles any preconditions for ignition.
+	Ignition(in IgnitionInput) ([]client.Object, error)
+
+	// ControlPlaneAvailable is called once cluster.Spec.ControlPlaneEndpoint.IsValid()
+	// returns true, typically after load balancers have been provisioned. It can be used
+	// to create DNS records.
+	ControlPlaneAvailable(in ControlPlaneAvailableInput) error
+}
+
+// PreProvisionInput collects the args to be passed to the PreProvision call.
+type PreProvisionInput struct {
+	ClusterID     string
+	InstallConfig *installconfig.InstallConfig
+}
```

Comment:
Within this step, we currently need to know (or be able to compute) whether masters are schedulable, to create appropriate security groups (CAPO can't do that yet and we may need to ship 4.16 like this).
We typically need the manifests for that, because the user may have changed the setting in the generated scheduler manifest.

WDYT?

Reply:
We would also use this hook to upload the RHCOS image to OpenStack. We used to do that as a side effect of the tfvars step.

If this is the right hook to use, then we would need to rhcosimage asset here. Or is there a better suited hook somewhere else?

Reply:
>Is there existing code that handles this in tfvars or does this not need to be handled in Terraform?

The Terraform script receives the value of `mastersSchedulable` from the Installer's generic code directly: 
https://github.com/openshift/installer/blob/0d56a06e02343e6603128e3f58c6c9bbc2edea3d/pkg/tfvars/tfvars.go#L59

Here is where it's computed:
https://github.com/openshift/installer/blob/0d56a06e02343e6603128e3f58c6c9bbc2edea3d/pkg/asset/cluster/tfvars.go#L170-L181

>We should consider passing tfvars into the CAPI provider

This is an interesting option. In [our CAPI development PR](https://github.com/openshift/installer/pull/7939) we are removing all side-effects from the OpenStack part of tfvars generation, so we should be good. We are moving all side effects to the `PreTerraform` hook of the Cluster step, which should then be guaranteed not to run within the CAPI flow.

Reply:
One alternative is to pass the entire `manifestsAsset` and we compute `mastersSchedulable` on the platform-specific implementation, potentially leaving more room for possible future needs.

Reply:
Shall I propose mastersSchedulable in PreProvisionInput as a separate PR?